src/day_2/ppo/solution/ppo.py

```diff
@@ -0,0 +1,273 @@
+"""PPO implementation for Pendulum-v1 environment.
+
+Reference:
+    - https://github.com/higgsfield/RL-Adventure-2/blob/master/3.ppo.ipynb
+"""
+
+from typing import Generator
+
+import gymnasium as gym
+import numpy as np
+import torch
+from icecream import ic
+from torch import nn, optim
+from torch.nn import functional as F
+from tqdm import trange
+
+USE_CUDA = torch.cuda.is_available()
+DEVICE = torch.device("cuda" if USE_CUDA else "cpu")
+
+
+class ActorNetwork(nn.Module):
+    """Actor Network which decide action."""
+
+    def __init__(self, state_dim: int, num_actions: int, hidden_size: int = 256):
+        super().__init__()
+
+        self.layers = nn.Sequential(
+            nn.Linear(state_dim, hidden_size),
+            nn.ReLU(),
+            nn.Linear(hidden_size, num_actions),
+        )
+
+        self.log_std = nn.Parameter(torch.ones(1, num_actions) * 0.0)
+
+    def forward(self, x: torch.Tensor) -> torch.distributions.Normal:
+        """Compute action distributions."""
+        mu = self.layers(x)
+        std = self.log_std.exp().expand_as(mu)
+        action_distribution = torch.distributions.Normal(mu, std)
+        return action_distribution
+
+
+class CriticNetwork(nn.Module):
+    """Critic Network which predict action value."""
+
+    def __init__(self, state_dim: int, hidden_size: int = 256):
+        super().__init__()
+
+        self.layers = nn.Sequential(
+            nn.Linear(state_dim, hidden_size),
+            nn.ReLU(),
+            nn.Linear(hidden_size, 1),
+        )
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Forward."""
+        return self.layers(x)
+
+
+class PPO:
+    """PPO method."""
+
+    def __init__(
+        self,
+        env_name: str,
+        num_envs: int = 1,
+        hidden_size: int = 256,
+        learning_rate: float = 3e-4,
+        num_steps: int = 2000000,
+        batch_size: int = 4096 * 4,
+        mini_batch_size: int = 256,
+        ppo_epochs: int = 4,
+        gamma: float = 0.99,
+        tau: float = 0.95,
+        clip_param: float = 0.2,
+    ) -> None:
+        # Create environments
+        self.env_name = env_name
+        self.env = gym.vector.make(self.env_name, num_envs=num_envs)
+        state_dim = self.env.observation_space.shape[1]  # type: ignore
+        num_actions = self.env.action_space.shape[1]  # type: ignore
```

Comment:
```suggestion
        state_dim = self.env.observation_space.shape[1]
        num_actions = self.env.action_space.shape[1]
```

linting이 없어서 빼고 될 것 같습니다.